### PR TITLE
DM-7225 File Upload Function for FITS, VOTable, IPAC, CSV, TSV files.

### PR DESCRIPTION
**Pull request by cwang2016**
The implementation includes: 
- a generic file upload function for files of FITS, VOTable, IPAC, CSV, TSV formats. 
  . for FITS format file, the fits could consist of one or more HDU units. 
    for the FITS with extensions, the extension could be either image or table type, or both. 
  . for VOTable file, the file could contain data of 'TABLEDATA', 'BINARY', 'BINARY2', or 'FITS' format. 

test: 
- localhost:8080/firefly
- select 'File Upload' tab
- upload file by clicking 'Choose File' 
   (there are some samples put under lsst/firefly_test_data/FileUpload-samples/CSV, TSV, IPAC, FITS, VOTABLE)
 - for votable and fits files, a summary table which reports the 'units' of the uploaded file is shown below the 'Choose File' button. 
    By highlighting the table row, the header of FITS HDU or the information of the table in VOTable is shown at the right of the summary table. 
    By selecting the table row(s) and clicking 'Search' button, the image or the table content will be further shown in the tri-view panel. 
- for CSV, TSV, IPAC files. the table content is rendered on the tri-view by clicking 'Search'

### CODE REVIEW

File: src/firefly/js/ui/FileUpload.jsx [hunk 1]
```diff
@@ -63,27 +63,27 @@ function getProps(params, fireValueChange) {
...
 
     fireValueChange({
         displayValue,
-        value: makeDoUpload(file, type)
+        value: !fileAnalysis ? makeDoUpload(file, type) : makeDoUpload(file, type, fileAnalysis)()
     });
 }
 
-function makeDoUpload(file, type) {
+function makeDoUpload(file, type, fileAnalysis) {
```
Comment (robyww) [inline src/firefly/js/ui/FileUpload.jsx]: When the file is uploading we need to mask the whole dialog.  Look at line 98 of ImageSelectDropdown.jsx`. 

Use `className='loading-mask'`

File: src/firefly/java/edu/caltech/ipac/util/FitsHDUUtil.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            dt.setShortDesc(meta.getDescription());
+            cols.add(dt);
+        }
+
+        DataGroup dg = new DataGroup("fits", cols);
+        String invalidMsg = "invalid fits file";
+
+        try {
+            Fits fits = new Fits(fitsFile); // open fits file
+            BasicHDU[] allHDUs = fits.read();
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/util/FitsHDUUtil.java]: Reading the full file to get the header info is wasteful and may take a long time for larger file.
Have you explore other options.. ie. nom.tam.fits.Header.readHeader?

File: src/firefly/java/edu/caltech/ipac/util/VoTableUtil.java [hunk 3]
```diff
@@ -45,6 +55,134 @@
...
+            DataType dt = new DataType(meta.getKey(), meta.getTitle(), meta.getMetaClass());
+            dt.setShortDesc(meta.getDescription());
+            cols.add(dt);
+        }
+        DataGroup dg = new DataGroup("votable", cols);
+        String invalidMsg = "invalid votable file";
+
+        try {
+            StarTableFactory stFactory = new StarTableFactory();
+            TableSequence tseq = stFactory.makeStarTables(voTableFile, null);
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/util/VoTableUtil.java]: Same as fits images.  Is there a way to get the headers only?  
Also, if we're reading the full file anyways, why not use existing function to read it into a DataGroup?
I am also concern with using json string as parsed values.  There's no way to reuse this code.

File: src/firefly/js/ui/FileUpload.jsx [hunk 4]
```diff
@@ -63,27 +63,27 @@ function getProps(params, fireValueChange) {
...
 
-function handleChange(ev, fireValueChange, type) {
+function handleChange(ev, fireValueChange, type, fileAnalysis) {
     var file = get(ev, 'target.files.0');
     var displayValue = get(ev, 'target.value');
 
     fireValueChange({
         displayValue,
-        value: makeDoUpload(file, type)
+        value: !fileAnalysis ? makeDoUpload(file, type) : makeDoUpload(file, type, fileAnalysis)()
```
Comment (loitly) [inline src/firefly/js/ui/FileUpload.jsx]: Is this check necessary?  if fileAnalysis is undefined, it'll be undefined regardless whether or not your pass it as a parameter.

**Comment by ejoliet** [comment]
Wow! That's great. 
I have noticed similar issues already mentioned by Trey and i'd like to add new one:
- why not using firefly-dev.html to add the "file upload" button, i don't think we are ready to put it in the top bar, are we? The ui on how to use the function to parse a file to be uploaded i don't think it was part of the ticket. There should be another ticket for designing the L&F of the usage after getting the information out of the file: format, summary, title, error message,etc.
- After uploading a FITS file, i've uploaded an XML file but the previous table from the previous upload stayed, i think it's a bug but again i don't know what the UI should look like.
- I have more comments on the panel itself but i feel this is not the right ticket/moment to discuss that.  
- (usre-friendly) Message error should be at a different layer and not coupled to this function

We should discuss further on how to use the improved function upload you are doing in FF apps in general.

File: src/firefly/js/visualize/ui/FileUploadViewPanel.jsx [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+                    const highlightInfo = get(analysisModel, ['tableMeta', tblIdx], '');
+                    const highlightObj = highlightInfo ? JSON.parse(highlightInfo) : undefined;
+                    const rowInfo = highlightObj && get(highlightObj, 'rowInfo');
+
+                    if (rowInfo) {
+                        const SP = '\u00a0';
+                        return rowInfo.map((entry) => {
+                            let   key = get(entry, 'key');
+                            const val = get(entry, 'value');
+                            const comment = get(entry, 'comment', '');
```
Comment (loitly) [inline src/firefly/js/visualize/ui/FileUploadViewPanel.jsx]: let {key, val, comment=''} = entry || {};
one-liner for the above 3 lines.

**Comment by loitly** [comment]
From UI test:
- when uploading a bad .fits file, JS errors appear in console but no indication in UI.  UI fail to recover, and will not work anymore.
- fail to load a sample lsst table fits file.
- on a very big file, no indication that it's working.  appears as if it did nothing.
- votable and fits file with only one entry should not appears in a table with selection box.

**Comment by ymeiymei** [comment]
Tested. Overall the file loader works great! 

Found some problems.

1)Load /hydra/cm/firefly_test_data/FileUpload-samples/fits/imagetable/calexp-005882-i1-0406.fits, load one image, then load one binary table. UI didn't show the table result and the error message in the Console is:
"utils.js:235 uncaught at masterSaga 
 at layoutManager 
 TypeError: Cannot read property 'zoomFactor' of null
    at http://127.0.0.1:8080/firefly/firefly-dev.js:117426:29
    at Object.dispatch (http://127.0.0.1:8080/firefly/firefly-dev.js:165908:19)
    at process (http://127.0.0.1:8080/firefly/firefly-dev.js:164700:16)
    at dispatchZoom (http://127.0.0.1:8080/firefly/firefly-dev.js:84323:6)
    at http://127.0.0.1:8080/firefly/firefly-dev.js:84435:18
    at Array.forEach (<anonymous>)
    at dispatchChangeExpandedMode (http://127.0.0.1:8080/firefly/firefly-dev.js:84432:32)
    at ExpandedModeDisplay.componentWillUnmount (http://127.0.0.1:8080/firefly/firefly-dev.js:184229:65)
    at http://127.0.0.1:8080/firefly/firefly-dev.js:43161:26
    at measureLifeCyclePerf (http://127.0.0.1:8080/firefly/firefly-dev.js:42827:13)"

If I load the binary table(s) first and then load image(s), it works fine. If I load images and binary tables at the same time, it works fine (I can see the result and UI has no error messages.)


2)Load /hydra/cm/firefly_test_data/FileUpload-samples/fits/imagetable/hpacs1342250905_00hps3d_00.fits. If you select/search all extensions, all the images won't be plotted. If you select/search one by one, you can get all but the history image. 
No error message in the console.

I expect no matter which way to select/search, I should get the ones with valid data.

Somehow hpacs1342250905_00hps3d_00.fits doesn't have the problem even the history image is not working too.

3) Failed to load 
/hydra/cm/firefly_test_data/FileUpload-samples/JSON/gaia_result.json
/hydra/cm/firefly_test_data/FileUpload-samples/VOTable/fits/gaia_fits.xml


4)No or very few metadata for 
/hydra/cm/firefly_test_data/FileUpload-samples/TSV/gaia_result.tsv
/hydra/cm/firefly_test_data/FileUpload-samples/VOTable/binary/* and binary2/*
/hydra/cm/firefly_test_data/FileUpload-samples/VOTable/fits/starlinkVOfits.xml
and more.

Design or bug?

5)My computer can't handle  select-all/search for /hydra/cm/firefly_test_data/FileUpload-samples/VOTable/tabledata/allpostables.xml. There are 164 tables. In the result page, I move around the mouse and things are messed up (see attached screenshot here).
loadtoomanytables

Cindy just told me she was still fixing the code. Maybe the issues above are already fixed. I'll continue testing while her new version is out.

**Comment by cwang2016** [comment]
Features and code updates:
-   Add the button for the user to select if rendering image extensions in one window or each image extension in separate window in case multiple image extensions are selected to search.
- Add support to read file from a URL as well as from file upload – two tabs under the upload dropdown panel are made from to do upload from file and URL respectively. 
-  Change ‘File Upload’ button to be ‘Upload’ from firefly.html
-  For FITS and VOTable meta data, showing the meta information in table style. For FITS file, it is the header of HDU, and it is non FIELD tag information attached to the table, like PARAM tag.
- For FITS and VOTable upload, two tables are shown as the file analysis results, one is for file summary and the other one contains the meta information on the highlighted table or image unit. 
- Add masking on the working panel during uploading and image/table searching.
- Show error messages for searching the table or image from the uploaded file in case there is no data contained, no file uploaded or no table/image unit (for FITS or VOTable) selected. 
- In default, the first table for VOTable or the primary HDU/the first extension for FITS which contains data is checked as selected from the summary table.
-  Set a number (currently 20) to limit the maximum table/image units to be searched for FITS or VOTable.
- Add more information into the description of the uploaded file, such as file size.
- Change the wording for IPAC and FITS.
- code changes per comments on the codes.

PS. gaia_fits.xml is a VOTable with wrong data.

**Comment by cwang2016** [comment]
@loitly  @robyww @ejoliet @ymeiymei 
There are more changes added. Please help review it again, especially some new features added to the UI.  Thanks.

**Comment by cwang2016** [comment]
Add some update: 
- the summary table selection and highlight are made to be in sync of table sorting. 

The following issues are planned to be solved by other tickets in the future
- table operation related issue: 
  Sorting a table of an loaded FITS or XML, the selected rows are set back to default. If i have 2 
  selected rows, it goes back to 1 
- file format identification issue:
   i've uploaded a wrong ASCII region files but it says "json not supported" (should probably say wrong format or something less precise than json i guess, similar to when i upload an HTML file)

**Comment by ymeiymei** [comment]
After Cindy updated the code, I tested again. Tried to load all the data files in firefly_test_data/FileUpload-samples. 

Most of them are loaded and shown in tri-view. Great job!!!

Found known issues:
No coverage map for objsearch.xml, multipleTables_Ned.xml, and stilbinary2.xml. Reason? 
No support on json files yet.
Some gaia data files have syntax problem so can't be loaded.

We can improve: 
When try to load more than ~500 images, firefly would fail. For example, load 
firefly_test_data/FileUpload-samples/fits/imagetable/1342179301_cubeHML_6b_1625_2016-01-21T19-59-34UTC.fits, select the extensions image, weight, flag and Search. It will fail.
We can put the image retrieving in the background?

Good job again!

**Comment by cwang2016** [comment]
the issues regarding FileUpload component styling will be investigated in more detail, and solved by another ticket if needed, 
- If an error is shown, the icon is overlapping the warning message.
- Seems that the tooltip is not correct (shows "C:\fakepath..."

**Comment by loitly** [comment]
Testing it again after multiple updates.  It errors out on a couple of known bad files.  Cindy is aware of it.  Other than that, it seem to work pretty well.
There are several cosmetic things I would like to point out.  You do not need to fix it.  Just my opinion.
- The use of tab panel just for File/URL selection is awkward.
- There should be a small gap between the 2 tables.
- The table's title can be improve
- Right table does not need highlighted row.  It has no use and could be a distraction.

**Comment by robyww** [comment]
It is looking good.  There is still some UI clean up I would like to see but all the functionality seems to all be there. I think you should merge it. We should then show it to a few people and make another ticket for any UI changes.

**Comment by cwang2016** [comment]
Some updates are made: 
- change the upload selection style from tabs to be a group of radio buttons 
- update the layout and the styling for the upload dropdown panel.

**Comment by cwang2016** [comment]
Note: 
this branch is meant to solve the issue for ticket DM-7225 instead of DM-7725.

**Pull request merged**